✨ [amp-social-share][bento] Handle children for social-share

krdwan opened the pull request:
Tracking Issue: https://github.com/ampproject/amphtml/issues/28283

Add functionality to conditionally display children in social-share

1. Use `passthrough` on amp-social-share and control display of children by including (or not) the unnamed slot
2. Create new attribute `custom-icon` on amp-social-share and `customIcon` prop on preact social-share
3. When `custom-icon` or `customIcon` is `true`, children will automatically be displayed by the Preact component
4. If `custom-icon` or `customIcon` is false, children will be displayed only when a `pre-configured` type is not provided

Preact rendered nodes will now be rendered in the shadow DOM (previously was rendered in light DOM)

**Design Changed to the following** -------

Updated, added a two config properties `canHaveChildren` and `canChangeColor` for each of the pre-configured `types`.  If these are false, children will automatically be hidden and default colors will be applied to the elements using inline styles.  I have defaulted all to `true` but will update if we hear back differently from p counsel.

When `canHaveChildren` is true, children will be displayed if they are provided and will override a pre-configured icon if it exists.

When `canChangeColor` is true, color and background color will be accepted from the props.  These will still default to the correct colors.  When `canChangeColor` is false, color and background color will have defaults applied directly and ignore props.

**A Note on !important:**
I have not been able to find a way of applying `!important` to Preact styles inline.  However, I think the use case for our styles being overwritten is pretty rare.  In amp-mode we are protected by shadow DOM encapsulation from what I understand.  In Preact mode, the user will need to overwrite the high-specificity inline style to change the colors which is already high effort.  So we may be able to get away without using this.  If I find an easy way to add this, I will create a new PR.

**A Note on Mutation Observer:**
Will add Mutation Observer to the component in a follow up PR.

caroqliu commented on extensions/amp-social-share/1.0/social-share.js:
```diff
@@ -57,6 +58,17 @@ export function SocialShare(props) {
   };
+
+  const icon =
+    !getSocialConfig(props['type']) || props['customIcon'] ? (
```
Is `customIcon` needed? Can you do `props['children'] || <SocialShareIcon ...>`?

krdwan replied:
From my understanding `props['children']` creates an unnamed slot in the shadow DOM (when using `passthrough`) so the children can be displayed, but actually does not represent those children.

And therefore from Preact side, there is actually no way of knowing whether or not the AMP side has provided children.

This is the basis for using `customIcon`, but if I'm not understanding correctly happy to be corrected!

dvoytenko replied:
@krdwan is correct that currently, the "passthrough" will pass the content as a single `slot` no matter whether or not there are children. But @caroqliu is correct that burdening the component with this nuance seems wrong. I think we just need to make `passthrough` smarter. The simplest thing we could for now is to create a "passthrough-non-empty" mode where we create a shadow DOM, but compute children as following:

```
props['children'] = this.getRealChildNodes().length > 0 ? [<Slot />] : null;
```

caroqliu commented on extensions/amp-social-share/1.0/amp-social-share.js:
```diff
@@ -93,6 +93,7 @@ class AmpSocialShare extends PreactBaseElement {
     return dict({
       'width': responsive || this.element.getAttribute('width'),
       'height': responsive || this.element.getAttribute('height'),
+      'customIcon': this.element.hasAttribute('custom-icon'),
```
This can be specified in `AmpSocialShare['props']`

krdwan replied:
Thanks, updated!

krdwan commented on extensions/amp-social-share/validator-amp-social-share.protoascii:
```diff
@@ -66,6 +66,9 @@ tags: {  # <amp-social-share>
     mandatory: true
   }
+  attrs: {
+    name: "custom-icon"
```
Thanks good to know.  Should the 1.0 validator be updated or documented somewhere else?

krdwan commented:
> IMO we shouldn't need to change the API surface this much in the AMP layer between 0.1 and 1.0. Discussed with @krdwan and IIUC there are two primary use-cases for `custom-icon`:
> 
> 1. Overriding a pre-configured icon, which can include
>    a. allowing users to modify default icons
>    b. allowing users to supply arbitrary children to render _instead of_ default icons
> 2. Providing an icon for a non-pre-configured type
> 
> In `0.1`, 1a and 2 are possible but 1b is not -- that is the component does not explicitly respect or prioritize user-supplied children for preconfigured provider types. This PR appears to make 1b and 2 possible but not necessarily 1a in `1.0`.
> 
> In Preact mode wouldn't these both be doable by rendering `props['children'] || <SocialShareIcon ...>`? In AMP mode we have [historically allowed people to do this to varying extents via CSS](https://amp.dev/documentation/examples/components/amp-social-share/?format=websites#custom-style). Have we confirmed that the rendered pre-configured content is not user-overridable via CSS in a similar fashion to the status quo?
> 
> Finally whether and to what extent to respect user-supplied children is not clear -- there are no validator rules against it but no existing samples on amp.dev to support it. I'm OK with adding `passthrough` but on the surface it seems unconventional that we'd need it.

For context, in 0.1, any children supplied by the user get rendered in addition to any icon that amp creates (they just end up overlapping).  I would imagine the user would only supply children when they are using a non-preconfigured type (`2` you mentioned from above).

krdwan commented:
Updated, added a two config properties `canHaveChildren` and `canChangeColor` for each of the pre-configured `types`.  If these are false, children will automatically be hidden and default colors will be applied to the elements using inline styles.  I have defaulted all to `true` but will update if we hear back differently from p counsel.

When `canHaveChildren` is true, children will be displayed if they are provided and will override a pre-configured icon if it exists.

When `canChangeColor` is true, color and background color will be accepted from the props.  These will still default to the same colors.  When `canChangeColor`is false, color and background color will have defaults applied via inline styles.

A Note on !important:
I have not been able to find a way of applying `!important` to Preact styles inline.  However, I think the use case for our styles being overwritten is pretty rare.  In amp-mode we are protected by shadow DOM encapsulation from what I understand.  In Preact mode, the user will need to overwrite the high-specificity inline style to change the colors which is already high effort.  So we may be able to get away without using this.  If I find an easy way to add this, I will create a new PR.

A Note on Mutation Obserber:
Will add Mutatino Observer to the component in a follow up PR.

dvoytenko commented on extensions/amp-social-share/1.0/social-share-config.js:
```diff
@@ -48,26 +48,34 @@ const BUILTINS = {
       'text': 'TITLE',
       'url': 'CANONICAL_URL',
     }),
+    'canChangeColor': true,
```
I think these will never be out of sync. So, we can just `canCustomize: boolean`.

krdwan replied:
Thanks, updated!

dvoytenko commented on extensions/amp-social-share/1.0/social-share.js:
```diff
@@ -74,12 +65,62 @@ export function SocialShare(props) {
+      'color':
+        typeConfig['canChangeColor'] && propsColor
+          ? propsColor
+          : CSS[uType]
```
At this point, having colors in `CSS[uType]` just makes our life harder. Let's move them to the `typeConfig`? The `CSS.BASE_STYLE` can stay where it is, of course.

krdwan replied:
Sounds good, updated!

dvoytenko commented on extensions/amp-social-share/1.0/social-share-config.js:
```diff
@@ -48,26 +48,38 @@ const BUILTINS = {
     'defaultParams': dict({
       'href': 'CANONICAL_URL',
     }),
+    'canCustomize': false,
```
Let's go ahead and set it to `false` for all branded icons. We can set back to `true` if we learn anything different later. But this is a more likely scenario right now, imho.

krdwan replied:
Sounds good, thanks!!

caroqliu replied:
I think this is always `true` now and therefore removable?

krdwan replied:
Thanks, in the latest version, we updated the company related one to `false`.  This way the user cannot change color or provide children for these `types`.  We can change it if we hear back differently from p counsel.

caroqliu commented on extensions/amp-social-share/1.0/amp-social-share.js:
```diff
@@ -158,6 +158,8 @@ class AmpSocialShare extends PreactBaseElement {
 /** @override */
 AmpSocialShare['Component'] = SocialShare;
 
+AmpSocialShare['passthroughNonEmpty'] = true;
```
Is this going to be defined in `PreactBaseElement`? Does it do anything now?

krdwan replied:
I added this in a separate PR here: https://github.com/ampproject/amphtml/pull/29367.  Wasn't really sure if I should document it somewhere.  I suppose the code base is the documentation for now.

Basically the behavior is the same as 'passthrough' except if there are no children it will return `null` instead of a `slot`.  And it's also recommended to use mutation observer when using this new `passthroughNonEmpty`.

caroqliu commented on extensions/amp-social-share/1.0/social-share.js:
```diff
@@ -74,12 +65,55 @@ export function SocialShare(props) {
+}
+
+/**
+ * If the specified type can have children and children exist return children.
+ * Otherwise return the default icon for the specified type.  If the specified
+ * type can have its color / background changed, set those to color /
+ * background specified in props, otherwise use defaults, or do not set if
+ * defaults do not exist.
```
nit: Can this be more clear? Something like:
```
Computes children based on four possible configurations in order:
  - Children provided
  - Color provided
  - Background color provided
  - Type provided

Otherwise, use icon defaults for all pre-configured types and empty <svg> otherwise.
```

Btw, I assume `backgroundColor` can apply and `color` is noop if provided for a non-configured type. Is that right?

krdwan replied:
Thanks, I have updated the comments to be a bit more clear.  For non-configured type, both `color` and `backgroundColor` can apply.  Although there won't really be anything for `color` to apply to unless the publisher provides children as well.
